File: Data Structures & Algorithms/path-with-maximum-probability/submission-3.kt
class Solution {
    fun maxProbability(n: Int, edges: Array<IntArray>, succProb: DoubleArray, start_node: Int, end_node: Int): Double {

        var adjList = HashMap<Double, MutableList<DoubleArray>>()
        var minHeap = PriorityQueue<DoubleArray>(Comparator<DoubleArray>() { c1, c2 -> 
            if (c1[1] < c2[1]) {
                1
            } else if (c1[1] > c2[1]) {
                -1
            } else {
                0
            }
        })

        for (i in 0 ..< n) {
            adjList[i.toDouble()] = mutableListOf()
        }

        var visited = HashSet<Double>()

        for (i in 0 until edges.count()) {
            var src = edges[i][0].toDouble()
            var dest = edges[i][1].toDouble()
            var prob = succProb[i]
            adjList[src]!!.add(doubleArrayOf(dest, prob))
            adjList[dest]!!.add(doubleArrayOf(src, prob))
        }
        minHeap.add(doubleArrayOf(start_node.toDouble(), 1.0))
        while (minHeap.count() > 0) {
            var node = minHeap.poll()!!
            var src = node[0]
            var probality = node[1]
            if (visited.contains(src)) {
                continue
            }
            if (src == end_node.toDouble()) {
                return probality
            }
            visited.add(src)
            var neighbors = adjList[src] ?: listOf()
            for (neighbor in neighbors) {
                minHeap.add(doubleArrayOf(neighbor[0], neighbor[1] * probality))
            }
        }
        return 0.0
    }
}

File: Data Structures & Algorithms/network-delay-time/submission-3.swift
class Solution {
    func networkDelayTime(_ times: [[Int]], _ n: Int, _ k: Int) -> Int {

        var adjList = [Int: [(Int, Int)]]()

        for time in times {
            let src = time[0]
            let dst = time[1]
            let time = time[2]
            if adjList[src] == nil { 
                adjList[src] = []
            }
            adjList[src]!.append((dst, time))
        }

        struct Node: Comparable {
            let dst: Int
            let time: Int

            static func < (n1: Node, n2: Node) -> Bool {
                return n1.time < n2.time
            }
        }

        var minHeap = Heap<Node>()
        minHeap.insert(Node(dst: k, time: 0))

        var visited = Set<Int>()

        while minHeap.count > 0 {
            let node = minHeap.popMin()!
            if visited.contains(node.dst) {
                continue
            }
            visited.insert(node.dst)
            if visited.count == n {
                return node.time
            }
            let neighbors = adjList[node.dst] ?? []
            for neighbor in neighbors {
                minHeap.insert(Node(dst: neighbor.0, time: node.time + neighbor.1))
            }
        }

        return -1

    }
}
